File: src/main/java/intricateengineers/intricatemachinery/api/module/ModuleItem.java
/*
 * Copyright (c) 2016 IntricateEngineers
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package intricateengineers.intricatemachinery.api.module;

import javax.vecmath.Vector3f;

import mcmultipart.multipart.IMultipart;
import mcmultipart.multipart.MultipartHelper;

import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.util.EnumActionResult;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;

public class ModuleItem extends Item {

    public final Class<? extends Module> module;

    public ModuleItem(Class<? extends Module> module) {
        this.module = module;
    }
    
    /**
     * Called when a Block is right-clicked with this Item
     */
    @Override
    public final EnumActionResult onItemUse(ItemStack stack, EntityPlayer playerIn, World worldIn, BlockPos pos, EnumHand hand, EnumFacing facing, float hitX, float hitY, float hitZ)
    {
        for (IMultipart part : MultipartHelper.getPartContainer(worldIn, pos).getParts()) {
            if (part instanceof MachineryFrame) {
                MachineryFrame frame = (MachineryFrame) part;
                //TODO: Raytracing. Let that function know which module was hit where
                if (this.placeInFrame(frame, stack, playerIn, hand, facing, new Vector3f(hitX, hitY, hitZ))) return EnumActionResult.SUCCESS; 
            }
        }

        return EnumActionResult.PASS;
    }

    /**
     * Do whatever you need to do when this item is rightclicked on a frame
     * @return whether the action was successfull
     * TODO: Raytracing. This function needs to know which module was clicked on which side.
     */
    public boolean placeInFrame(MachineryFrame frame, ItemStack stack, EntityPlayer player, EnumHand hand, EnumFacing facing, Vector3f hit) {
        //TODO: Place at correct positions
        try {
            frame.addModule(module.newInstance());
            return true;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

}

File: src/main/scala/intricateengineers/intricatemachinery/api/module/ModuleItem.scala
package intricateengineers.intricatemachinery.api.module

import intricateengineers.intricatemachinery.api.model.Box
import mcmultipart.multipart.{IMultipartContainer, MultipartHelper}
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.init.SoundEvents
import net.minecraft.item.{EnumAction, Item, ItemStack}
import net.minecraft.util.{EnumActionResult, _}
import net.minecraft.util.math.{AxisAlignedBB, BlockPos}
import net.minecraft.world.World
import org.lwjgl.util.vector.Vector3f

import scala.collection.JavaConversions._

class ModuleItem[T <: ModuleCompanion](val moduleObject: T, val createModule: (MachineryFrame) => Module) extends Item {
  final val name = moduleObject.Name

  override final def onItemUse(stack: ItemStack, playerIn: EntityPlayer, worldIn: World, pos: BlockPos,
                               hand: EnumHand, facing: EnumFacing, hitX: Float, hitY: Float, hitZ: Float): EnumActionResult = {

    // BlockPos facing the block that was right clicked
    val posFacing = pos.add(facing.getDirectionVec)

    val container = Option(MultipartHelper.getPartContainer(worldIn, pos))
    val containerFacing = Option(MultipartHelper.getPartContainer(worldIn, posFacing))

    val hitVec = new Vector3f(hitX, hitY, hitZ)

    // Was the player looking at the MF itself or a Module inside it?
    isContainerAMachineryFrame(container).map(maybeFrame => {
      if(placeInFrame(maybeFrame, playerIn, worldIn, pos, facing, hitVec, isFromBlockFace = false))
        EnumActionResult.SUCCESS
      else
        EnumActionResult.FAIL
    }
    )

    // Was the player looking at a Block next to the MF?
    isContainerAMachineryFrame(containerFacing).map(maybeFrame => {
      if (placeInFrame(maybeFrame, playerIn, worldIn, pos, facing, hitVec, isFromBlockFace = true))
        EnumActionResult.SUCCESS
      else
        EnumActionResult.FAIL
    }
    )

    EnumActionResult.PASS
  }

  // Place when the player was looking at a Block next to the MF
  @inline
  private def placeInFrame(frame: MachineryFrame, playerIn: EntityPlayer, worldIn: World, pos: BlockPos,
                           facing: EnumFacing, hit: Vector3f, isFromBlockFace: Boolean): Boolean = {

    val newModule = createModule(frame)
    val mainBox = newModule.model.mainBox
    val moduleSizeNormalized = normalizeModuleSize(mainBox)

    val modulePosVec: Vector3f = hit

    if (isFromBlockFace) {
      facing match {
        case EnumFacing.NORTH =>
          modulePosVec.setZ(1 - modulePosVec.z - moduleSizeNormalized.z)
        case EnumFacing.WEST =>
          modulePosVec.setX(1 - modulePosVec.x - moduleSizeNormalized.x)
        case EnumFacing.DOWN =>
          modulePosVec.setY(1 - modulePosVec.y - moduleSizeNormalized.y)
        case _ =>
      }
    }
    else {
      facing match {
        case EnumFacing.NORTH =>
          modulePosVec.setZ(modulePosVec.z - moduleSizeNormalized.z)
        case EnumFacing.WEST =>
          modulePosVec.setX(modulePosVec.x - moduleSizeNormalized.x)
        case EnumFacing.DOWN =>
          modulePosVec.setY(modulePosVec.y - moduleSizeNormalized.y)
        case _ =>
      }
    }

    facing.getAxis match {
      case EnumFacing.Axis.X =>
        modulePosVec.setZ((modulePosVec.z - moduleSizeNormalized.z / 2).max(0))
        modulePosVec.setY((modulePosVec.y - moduleSizeNormalized.y / 2).max(0))
      case EnumFacing.Axis.Z =>
        modulePosVec.setX((modulePosVec.x - moduleSizeNormalized.x / 2).max(0))
        modulePosVec.setY((modulePosVec.y - moduleSizeNormalized.y / 2).max(0))
      case EnumFacing.Axis.Y =>
        modulePosVec.setZ((modulePosVec.z - moduleSizeNormalized.z / 2).max(0))
        modulePosVec.setX((modulePosVec.x - moduleSizeNormalized.x / 2).max(0))
    }
    newModule.pos = correctBounds(modulePosVec, moduleSizeNormalized)

    if(isModuleObstructed(frame, newModule)) {
      return false
    }

    frame.modules += newModule

    playerIn.swingArm(EnumHand.MAIN_HAND)
    worldIn.playSound(playerIn, pos, SoundEvents.BLOCK_STONE_PLACE, SoundCategory.BLOCKS, 1.0F, 1.0F)
    return true
  }

  private def correctBounds(pos: Vector3f, boxSize: Vector3f): ModulePos = {
    ModulePos(
      limit(pos.x % 1, 0f, 1f - boxSize.x),
      limit(pos.y % 1, 0f, 1f - boxSize.y),
      limit(pos.z % 1, 0f, 1f - boxSize.z)
    )
  }

  def isModuleObstructed(frame: MachineryFrame, newModule: Module): Boolean = {
    val moduleAABB = newModule.model.mainBox.aabb.offset(newModule.pos.dX, newModule.pos.dY, newModule.pos.dZ)
    println(moduleAABB)
    //frame.bbCache().foreach(println)
    frame.bbCache().exists(_.intersectsWith(moduleAABB))

  }

  private def normalizeModuleSize(moduleBox: Box) = {
    new Vector3f(moduleBox.size.x / Module.GRID_SIZE, moduleBox.size.y / Module.GRID_SIZE, moduleBox.size.z / Module.GRID_SIZE)
  }

  def limit(x: Double, min: Double, max: Double): Double =
    if (x < min) min else if (x > max) max else x

  override def onEntitySwing(entityLiving: EntityLivingBase, stack: ItemStack): Boolean = false

  override def getItemUseAction(stack: ItemStack): EnumAction = EnumAction.BLOCK

  def isContainerAMachineryFrame(iMultipartContainer: Option[IMultipartContainer]): Option[MachineryFrame] = {
    iMultipartContainer
            .map(_.getParts.find(_.isInstanceOf[MachineryFrame]))   // Search for an MF in the parts of the container
            .map(_.get.asInstanceOf[MachineryFrame])   // If we found an MF, cast it to one and return it
  }
}
